File: cpp/egrabber-sample-programs/cpp/egrabber-snippets/samples/100-grabn.cpp
#include "../tools/tools.h"

#include <EGrabber.h>
#include <FormatConverter.h>

using namespace Euresys;

static void sample() {
    EGenTL genTL(Coaxlink()); // load CoaxLink GenTL producer
    EGrabberDiscovery egrabberDiscovery(genTL);
    egrabberDiscovery.discover(false); // grabber-oriented discovery
    EGrabber<CallbackOnDemand> grabber(egrabberDiscovery.egrabbers(0));
    FormatConverter converter(genTL); // create rgb converter environment

    grabber.reallocBuffers(20); // prepare 20 buffers

    grabber.start(20); // grab 20 buffers
    for (size_t frame = 0; frame < 20; ++frame) {
        ScopedBuffer buffer(grabber); // wait and get a buffer
        // Note: ScopedBuffer pushes the buffer back to the input queue automatically
        BufferInfo bi = buffer.getInfo();
        std::string outFormat = ".jpeg";
        if (bi.pixelFormat == "Data8" || bi.width == 0) {
            outFormat = ".raw";
        }
        std::string name = Tools::getEnv("sample-output-path") + "/frame.NNN" + outFormat;
        buffer.saveToDisk(name, frame);
    }
}

static Tools::Sample addSample(__FILE__, sample, "Simple Grab N frames using ScopedBuffer class");

File: cpp/egrabber-sample-programs/python/100-grabn.py
"""Simple Grab N using 'with Buffer'"""

import os
outdir = '{}/{}.output'.format(os.getenv('OUTPUT_DIR', os.path.dirname(__file__)), os.path.splitext(os.path.basename(__file__))[0])
if not os.path.isdir(outdir):
    os.makedirs(outdir)

from egrabber import *

gentl = EGenTL()
grabber = EGrabber(gentl)

N = 5
grabber.realloc_buffers(N)

grabber.start(N)
for frame in range(N):
    with Buffer(grabber, timeout=1000) as buffer:
        # Note: the buffer will be pushed back to the input queue automatically
        # when execution of the with-block is finished
        rgb = buffer.convert('RGB8')
        rgb.save_to_disk(outdir + '/frame.{}.jpeg'.format(frame))
